task02-grep/src/main/java/ru/xamgore/tasks/commands/Grep.java

```
@@ -0,0 +1,115 @@
+package ru.xamgore.tasks.commands;
+
+import ru.xamgore.tasks.Task;
+
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.text.NumberFormat;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import org.apache.commons.cli.*;
+
+import static java.lang.Integer.parseUnsignedInt;
+
+
+/**
+ * Grep is a program to search text.
+ */
+public class Grep extends Task {
+  private static final String IGNORE_CASE_OPTION = "i";
+  private static final String WHOLE_WORDS_OPTION = "w";
+  private static final String AFTER_LINES_OPTION = "A";
+
+  private CommandLine cmd;
+  private Pattern regex;
+  private List<String> lines;
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public int exec() {
+    stdout = "";
+
+    try {
+      parseArgs();
+
+      // check file exists & get all lines, or take stdin
+      String fileArg = cmd.getArgs()[0];
+      lines = cmd.getArgs().length > 1
+        ? Files.readAllLines(Paths.get(fileArg))
+        : Arrays.asList(stdin.split("\\r?\\n"));
+
+      // make a regex pattern from user's input
+      String ignore_case = cmd.hasOption(IGNORE_CASE_OPTION) ? "(?i)" : "";
+      String whole_words = cmd.hasOption(WHOLE_WORDS_OPTION) ? "\\b" : "";
+
+      // the last argument is a regex
+      String user_pattern = cmd.getArgs()[cmd.getArgs().length - 1];
+
+      // populate regex with additional options
+      regex = Pattern.compile(ignore_case + "(" + whole_words + user_pattern + whole_words + ")");
+
+      // parse -A argument
+      Integer lines_after = parseUnsignedInt(cmd.getOptionValue(AFTER_LINES_OPTION, "0"));
+
+      stdout = process(lines_after);
+    } catch (NumberFormatException e) {
+      System.out.println("-A argument requires positive integer");
+      throw e;
+    } catch (Exception e) {
```

Лучше не ловить незнакомые исключения. Fail fast же.

---

Тогда упадёт весь шелл. От внутренних команд требуется, чтобы они всегда завершались и возвращали код возврата.

---

Мне кажется, это лучше делать в основном цикле